smart-py/objkt_swap_v3_test.py

```diff
@@ -531,3 +541,661 @@ def test_pause_contract():
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # try cancel someone elses as admin
+    scenario += marketplaceV3.cancel_swap(0).run(
+        sender = admin,
+        valid = False
+    )
+
+    # first one still exists
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # cancel own swap successfully
+    scenario += marketplaceV3.cancel_swap(0).run(
+        sender = artist1,
+        valid = True
+    )
+
+    # swap is gone
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == False)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+
+    # but id still incremented
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+@sp.add_test(name="Test collect v3 swap failure conditions")
+def test_collect_swap_failure_conditions():
+    # Get the test environment
+    testEnvironment = get_test_environment()
+    scenario = testEnvironment["scenario"]
+    admin = testEnvironment["admin"]
+    artist1 = testEnvironment["artist1"]
+    artist2 = testEnvironment["artist2"]
+    collector1 = testEnvironment["collector1"]
+    collector2 = testEnvironment["collector2"]
+    objkt = testEnvironment["objkt"]
+    marketplaceV1 = testEnvironment["marketplaceV1"]
+    marketplaceV3 = testEnvironment["marketplaceV3"]
+
+    # Mint an OBJKT
+    scenario += marketplaceV1.mint_OBJKT(
+        address=artist1.address,
+        amount=1,
+        metadata=sp.pack("ipfs://fff"),
+        royalties=100
+    ).run(
+        sender=artist1
+    )
+
+    # Artist updates the objkt to be managed by marketplace v3
+    scenario += objkt.update_operators(
+        [sp.variant("add_operator", objkt.operator_param.make(
+            owner=artist1.address,
+            operator=marketplaceV3.address,
+            token_id=152))]).run(sender=artist1)
+
+    # Successfully swap
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=1,
+        xtz_per_objkt=sp.mutez(1),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        sender=artist1,
+        valid=True
+    )
+
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # try to collect nonexistent swap and fail
+    scenario += marketplaceV3.collect(100).run(
+        sender = collector1,
+        amount = sp.mutez(1),
+        valid = False
+    )
+
+    # first one still exists
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # try to collect own swap and fail
+    #
+    # TODO should not be able to collect own swap
```

Comment:
@jagracar this test should fail, it should not be possible to collect your own swap

Reply:
Ok, i will add a verify sp.sender != swap.issuer